src/coins.cpp

```diff
@@ -158,35 +188,85 @@ bool CCoinsViewCache::GetNullifier(const uint256 &nullifier, NullifierType type)
     return tmp;
 }
 
-void CCoinsViewCache::PushAnchor(const ZCIncrementalMerkleTree &tree) {
+template<typename Tree, typename Cache, typename CacheIterator, typename CacheEntry>
+void CCoinsViewCache::AbstractPushAnchor(
+    const Tree &tree,
+    ShieldedType type,
+    Cache &cacheAnchors,
+    uint256 &hash
+)
+{
     uint256 newrt = tree.root();
 
-    auto currentRoot = GetBestAnchor();
+    auto currentRoot = GetBestAnchor(type);
 
     // We don't want to overwrite an anchor we already have.
-    // This occurs when a block doesn't modify mapAnchors at all,
+    // This occurs when a block doesn't modify mapSproutAnchors at all,
     // because there are no joinsplits. We could get around this a
-    // different way (make all blocks modify mapAnchors somehow)
+    // different way (make all blocks modify mapSproutAnchors somehow)
     // but this is simpler to reason about.
     if (currentRoot != newrt) {
-        auto insertRet = cacheAnchors.insert(std::make_pair(newrt, CAnchorsCacheEntry()));
-        CAnchorsMap::iterator ret = insertRet.first;
+        auto insertRet = cacheAnchors.insert(std::make_pair(newrt, CacheEntry()));
+        CacheIterator ret = insertRet.first;
 
         ret->second.entered = true;
         ret->second.tree = tree;
-        ret->second.flags = CAnchorsCacheEntry::DIRTY;
+        ret->second.flags = CacheEntry::DIRTY;
 
         if (insertRet.second) {
             // An insert took place
             cachedCoinsUsage += ret->second.tree.DynamicMemoryUsage();
         }
 
-        hashAnchor = newrt;
+        hash = newrt;
     }
 }
 
-void CCoinsViewCache::PopAnchor(const uint256 &newrt) {
-    auto currentRoot = GetBestAnchor();
+void CCoinsViewCache::PushSproutAnchor(const ZCIncrementalMerkleTree &tree) {
+    AbstractPushAnchor<ZCIncrementalMerkleTree, CAnchorsSproutMap, CAnchorsSproutMap::iterator, CAnchorsSproutCacheEntry>(
```

Comment:
Why do we have two methods rather than follow the same form as PopAnchor/AbstractPopAnchor?

Reply:
Virtual functions. I would have preferred to do that and tried to do it as much as possible.

Reply:
I figured it was something like this.